packages/edge-preview-authenticated-proxy/src/index.ts

```diff
@@ -0,0 +1,328 @@
+			{
+				error: this.name,
+				message: this.message,
+			},
+			{ status: this.status }
+		);
+	}
+
+	get data(): Record<string, unknown> {
+		return {};
+	}
+}
+
+class NoExchangeUrl extends HttpError {
+	constructor() {
+		super("No exchange_url provided", 400);
+	}
+}
+class ExchangeFailed extends HttpError {
+	constructor(
+		readonly url: string,
+		readonly exchangeStatus: number,
+		readonly body: string
+	) {
+		super("Exchange failed", 400);
+	}
+
+	get data(): { url: string; status: number; body: string } {
+		return { url: this.url, status: this.exchangeStatus, body: this.body };
+	}
+}
+class TokenUpdateFailed extends HttpError {
+	constructor() {
+		super("Provide token, prewarmUrl and remote", 400);
+	}
+}
+class RawHttpFailed extends HttpError {
+	constructor() {
+		super("Provide token, and remote", 400);
+	}
+}
+class PreviewRequestFailed extends HttpError {
+	constructor() {
+		super("Provide token, and remote", 400);
+	}
+}
+
+async function handleRawHttp(request: Request, url: URL) {
+	if (request.method === "OPTIONS") {
+		return new Response(null, {
+			headers: {
+				"Access-Control-Allow-Origin": request.headers.get("Origin") ?? "",
+				"Access-Control-Allow-Method": "*",
+				"Access-Control-Allow-Credentials": "true",
+				"Access-Control-Allow-Headers": "x-cf-token,x-cf-remote",
+				"Access-Control-Expose-Headers": "*",
+				Vary: "Origin",
+			},
+		});
+	}
+	const token = request.headers.get("X-CF-Token");
+	const remote = request.headers.get("X-CF-Remote");
+	if (!token || !remote) {
+		throw new RawHttpFailed();
+	}
+	const workerUrl = new URL(url);
+	const remoteUrl = new URL(remote);
+	workerUrl.hostname = remoteUrl.hostname;
+	workerUrl.protocol = remoteUrl.protocol;
+
+	const workerResponse = await fetch(workerUrl, {
+		...request,
+		headers: {
+			...request.headers,
+			"cf-workers-preview-token": token,
+		},
+	});
+	// The client needs the raw headers from the worker
+	// Prefix them with `cf-ew-raw-`, so that response headers from _this_ worker don't interfere
+	const rawHeaders: Record<string, string> = {};
+	for (const header of workerResponse.headers.entries()) {
+		rawHeaders[`cf-ew-raw-${header[0]}`] = header[1];
+	}
+	return new Response(workerResponse.body, {
+		...workerResponse,
+		headers: {
+			...rawHeaders,
+			"Access-Control-Allow-Origin": request.headers.get("Origin") ?? "",
+			"Access-Control-Allow-Method": "*",
+			"Access-Control-Allow-Credentials": "true",
+			"Access-Control-Allow-Headers": "x-cf-token,x-cf-remote",
+			"cf-ew-status": workerResponse.status.toString(),
+			"Access-Control-Expose-Headers": "*",
+			Vary: "Origin",
+		},
+	});
+}
+
+async function updatePreviewToken(url: URL, ctx: ExecutionContext) {
+	const token = url.searchParams.get("token");
+	const prewarmUrl = url.searchParams.get("prewarm");
+	const remote = url.searchParams.get("remote");
+	if (!token || !prewarmUrl || !remote) {
+		throw new TokenUpdateFailed();
+	}
+	if (prewarmUrl && token) {
```

Comment:
According to the if statement above, this must always be true, right?
Should `prewarmUrl` actually be optional?